### DISCUSSION TITLE
Authentication and whitelisting

### DISCUSSION THREAD

**Discussion by amasotti** [api-platform/api-platform#1972] [Q&A]
Hi, 

I'm implementing an API for my company and I've decided to try out api-platform, that seems fantastic to me. The main project is a [Pimcore](https://github.com/pimcore/pimcore) (based on Symfony 4) application. I've installed api-platform and managed to expose some entities, register some operations and serializations option.. no problem so far.

I'm however stuck on the authentication. I would avoid installing the JWT bundle if possible, since only very few user will have access to this API and they'll receive a token or apikey from us.
This token will be passed in the request header and be used for two main puroposes:
* Authorize the access to the API
* Filter which data will be provided: The token is associated with an instance of an "ApiUser" Object in the Db, which also contains many-to-may relations to some instances of exposed entities, to be used as a whitelist of the instances they're allowed to access or modify).

What's the best way to implement this in api-platform? I was thinking about the following options:
* If I leverage the Event System, I have the problem that I can authorize or deny the access but how do I pass the retrieved ApiUser whitelist to the DataProvider, so that I can restrict the SQL Query? 
* Or can I implement by DataProvider so that it extends a Symfony Controller and then use an onKernelEvent there?
* Separate Authentication and Filtering, but how do I have access to the Request header in the DataProvider?

Are there other, better options to do this? Or I'm thinking in the wrong direction?

**Comment by ngrafa** [comment]
Hi,

You can do this by writing a custom authenticator as described here. The first example near the top of this page should be sufficient. Don't forget to enable your authenticator in `config/packages/security.yaml`

https://symfony.com/doc/current/security/custom_authenticator.html

No need to use JWT bundle if you are issuing and governing the API key yourself.

-- Raf.
